daemon/api_sideload_n_try.go

```diff
@@ -207,6 +205,105 @@ out:
 	return AsyncResponse(nil, chg.ID())
 }
 
+// maxReadBuflen is the maximum buffer size for reading the non-file parts in the snap upload form
+const maxReadBuflen = 1024 * 1024
+
+// readForm returns a Form populated with values (for non-file parts) and file headers (for file
+// parts). The file headers contain the original file name and a path to the persisted file in
+// dirs.SnapDirBlob. If an error occurs and a non-nil Response is returned, an attempt is made
+// to remove temp files.
+func readForm(reader *multipart.Reader) (_ *Form, apiErr *apiError) {
+	maxMemory := int64(maxReadBuflen)
+	form := &Form{
+		Value:      make(map[string][]string),
+		FileHeader: make(map[string][]*FileHeader),
+	}
+
+	// clean up if we're failing the request
+	defer func() {
+		if apiErr != nil {
+			form.RemoveAll()
+		}
+	}()
+
+	for {
+		part, err := reader.NextPart()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return nil, BadRequest("cannot read POST form: %v", err)
+		}
+
+		name := part.FormName()
+		if name == "" {
+			continue
+		}
+
+		filename := part.FileName()
+		if filename == "" {
+			// non-file parts are kept in memory
+			buf := &bytes.Buffer{}
+
+			// copy one byte more than the max so we know if it exceeds the limit
+			n, err := io.CopyN(buf, part, maxMemory+1)
+			if err != nil && !errors.Is(err, io.EOF) {
+				return nil, BadRequest("cannot read form data: %v", err)
+			}
+
+			maxMemory -= n
+			if maxMemory < 0 {
+				return nil, BadRequest("cannot read form data: exceeds memory limit")
+			}
+
+			form.Value[name] = append(form.Value[name], buf.String())
+			continue
+		}
+
+		// file parts are persisted
+		tmpf, err := ioutil.TempFile(dirs.SnapBlobDir, dirs.LocalInstallBlobTempPrefix)
+		if err != nil {
+			return nil, InternalError("cannot create temp file for form data file part: %v", err)
+		}
+		defer func() {
+			if cerr := tmpf.Close(); err == nil && cerr != nil {
+				err = InternalError("cannot close temp file: %v", cerr)
+			}
+		}()
```

Comment:
Correct me if I'm wrong but a close doesn't imply a sync, so we would always need that call. From the close doc, "A successful close does not guarantee that the data has been successfully saved to disk (...) use fsync(2)". 
Regarding the open files, I'm not sure what's best tbh because we want to close in case of success, copy failure and sync failure so not deferring the close would require us to triplicate it

Reply:
What about a separate function (or an anonymous one, if you prefer) that takes a `part` as parameter and returns the file path where the part has been written to? Then you could use `defer` and still clean up the file as soon as it's ready.

Because otherwise in the (very unlikely) event that everything succeeds but that one of the files fails to close, this function will return an error but leave all the files in place.

Reply:
I'm not sure I understand your last point, because we try to remove all the files (in the first deferred function) if there's an error. In any case, I refactored the persisting to a separate function like you mentioned in the latest commit

Reply:
You are right, `RemoveAll` was actually handling it. Anyway, thanks for the update!